internal/catalogmetadata/client/client_test.go

```diff
@@ -60,59 +58,51 @@ func TestClient(t *testing.T) {
 										"name": "fake1.v9.9.9"
 									}
 								]
-							}`),
-						},
-					})
+							}`)...)
 
-					return objs, nil
+					return objs, nil, catalogContentMap
 				},
 				wantErr: `bundle "fake1.v9.9.9" not found in catalog "catalog-1" (package "fake1", channel "channel-with-missing-bundle")`,
 			},
 			{
 				name: "invalid bundle",
-				fakeCatalog: func() ([]client.Object, []*catalogmetadata.Bundle) {
-					objs, _ := defaultFakeCatalog()
+				fakeCatalog: func() ([]client.Object, []*catalogmetadata.Bundle, map[string][]byte) {
+					objs, _, catalogContentMap := defaultFakeCatalog()
 
-					objs = append(objs, &catalogd.CatalogMetadata{
-						ObjectMeta: metav1.ObjectMeta{
-							Name:   "catalog-1-broken-bundle",
-							Labels: map[string]string{"schema": declcfg.SchemaBundle, "catalog": "catalog-1"},
-						},
-						Spec: catalogd.CatalogMetadataSpec{
-							Content: json.RawMessage(`{"name":123123123}`),
-						},
-					})
+					catalogContentMap["catalog-1"] = append(catalogContentMap["catalog-1"], []byte(`{"schema": "olm.bundle", "name":123123123}`)...)
 
-					return objs, nil
+					return objs, nil, catalogContentMap
 				},
-				wantErr: "error unmarshalling catalog metadata: json: cannot unmarshal number into Go struct field Bundle.name of type string",
+				wantErr: "error processing response: expected value for key \"name\" to be a string, got %!t(float64=1.23123123e+08): 1.23123123e+08",
```

Comment:
I think this refactoring changes the meaning of the test case and probably should be a separate test case.

After refactoring we are hitting this:

```go
if err != nil {
	return err
}
```
in the very beginning of `walkFn` we pass into `WalkMetasReader`.

This means that `invalid bundle` test case and `invalid channel` test case below are basically testing the same code path.

The original was testing output of `Bundles` method when we had a overall valid looking response from the API, but marshalled bundle was invalid (or channel in case of `invalid channel` test case).

![Screenshot 2023-09-18 at 10 30 56](https://github.com/operator-framework/operator-controller/assets/509198/64c6f866-5279-4fc3-89d1-659195df9efe)

Reply:
Ah, good catch! I think this has to do with the actual unmarshalling behavior. It looks like if the response can't be unmarshalled into a `declcfg.Meta` object the first error case is triggered. I'll add some additional testing logic here to make sure we hit all the paths

Reply:
Updated the test cases in https://github.com/operator-framework/operator-controller/pull/411/commits/79cf2386f67dfe6578dcf757139dde69385c748d